jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/PathResourceTest.java

```diff
@@ -236,6 +236,46 @@ public void testGetPathTo(WorkDir workDir) throws IOException
         }
     }
 
+    @Test
+    public void testJarMountNonExistent(WorkDir workDir) throws IOException
+    {
+        Path tmpPath = workDir.getEmptyPathDir();
+        Path testJar = tmpPath.resolve("test.jar");
+
+        Map<String, String> env = new HashMap<>();
+        env.put("create", "true");
+
+        URI jarUri = URIUtil.uriJarPrefix(testJar.toUri(), "!/");
+        try (FileSystem zipfs = FileSystems.newFileSystem(jarUri, env))
+        {
+            Path root = zipfs.getPath("/");
+            Files.writeString(root.resolve("test.txt"), "Contents of test.txt", StandardCharsets.UTF_8);
+
+            Path dir = root.resolve("datainf");
+            Files.createDirectory(dir);
+            Files.writeString(dir.resolve("info.txt"), "Contents of info.txt", StandardCharsets.UTF_8);
+        }
+
+        try (ResourceFactory.Closeable resourceFactory = ResourceFactory.closeable())
+        {
+            // First reference to Resource in test.jar
+            Resource resRoot = resourceFactory.newResource(jarUri);
+            Resource resBadDir = resourceFactory.newResource(jarUri.toASCIIString() + "does-not-exist/");
+            assertNull(resBadDir);
+            Resource resBadFile = resourceFactory.newResource(jarUri.toASCIIString() + "bad/file.txt");
+            assertNull(resBadFile);
+            // Second reference to Resource in test.jar
+            Resource resInfoTxt = resourceFactory.newResource(jarUri.toASCIIString() + "datainf/info.txt");
+            assertTrue(Resources.isReadableFile(resInfoTxt));
+
+            if (resourceFactory instanceof ResourceFactoryInternals.Mountable mountable)
+            {
+                List<FileSystemPool.Mount> mounts = mountable.getMounts();
+                assertThat(mounts.size(), is(2));
```

Comment:
@gregw cannot have this be 1.

As the FileSystemPool is JVM wide, not ResourceFactory specific.
If 2 Resources, in different ResourceFactories access the same JAR these need to be tracked via the Resource -> Mount so that they can be dereferenced and FileSystem closed at the right time.

If we have `resRoot` cause the mount, and `resInfoTxt` reuse the mount, then there's not sufficient information to know how to eventually close that mount (especially so when the same jar is referenced as a Resource in 2 different ResourceFactory instances)

Reply:
@joakime if two `ResourceFactory`s open the same jar file, then each should have it listed as a single mount point. When both factories are closed, then the mount point can be unmounted.    How many resources that a given factory creates against its single mount is irrelevant to the reference counting for the close.

Fundamentally, the number of mount points in the factory for:
```java
   Resource foo = factory.newResource("jar:file:///tmp/test.jar!/foo/");
   Resource bar = factory.newResource("jar:file:///tmp/test.jar!/foo/bar.txt");
```
should be identical to 

```java
   Resource foo = factory.newResource("jar:file:///tmp/test.jar!/foo/");
   Resource bar = foo.resolve("bar.txt");
```

The count of mounts must be independent of how many resources are created against it.

The problem is in `mountIfNeeded(URI uri)` as it is always mounting, even if not needed.     It should be checking existing mounts and not mount again if it is already mounted.